File: client/sources/ci.py
"""Whether the default branch of each desk's repo actually builds.

The runner opens PRs into these repos all day. A red default branch is the most
expensive thing not to notice, because nothing stops: the PRs keep stacking up
and every one of them is built on a broken base.

WHY THIS IS ONE QUERY AND NOT SEVENTY
-------------------------------------
Seventy repos times one REST call each, every ten minutes, is four hundred and
twenty calls an hour and a rate limit by lunchtime. So the head of each default
branch is asked for through GraphQL, aliased, `CHUNK` repos at a time: seventy
repos is two calls per push. The answer is then cached on disk for `ttl()`
seconds, so a push that comes round early spends nothing at all.

Because it is cached, the payload carries `fetched_at` and every row carries
`checked_at`. A green light with no age on it is the false-green this project
exists to kill.

`checked_at` comes from the check runs themselves and is null when none of them
reported a completion time. The commit's own date travels separately as
`commit_at` and is NEVER substituted: a commit time shown as a check time is an
estimate with a measurement's face on.

THE STATES, AND WHY THERE ARE SIX
---------------------------------
  passing  the rollup says success
  failing  the rollup says failure or error. The alarm
  never    there are workflow files, and the head commit has no rollup at all.
           Usually a workflow that is filtered out of every branch it should
           run on, which is a different afternoon from a build that broke
  none     no workflow files. A repo with no CI is a DECISION, never a fault,
           and dressing it as one is how a docs repo goes permanently red
  running  the rollup is pending. Folding an in-flight run into green or red
           is a lie either way, and it is the one that resolves itself
  unknown  the repo could not be read: no token we hold can see it, or GraphQL
           refused. "We did not look" must never render as "it is fine"

This runs inside the snapshot push on the laptop, which is the only place a
GitHub credential exists. It can never move into the browser.
"""

from __future__ import annotations

import json
import os
import pathlib
import subprocess
import time

KEY = "ci"

# sections.read_all() hands the desk list to any source that asks for it. This
# is the only source that needs to know what the room is showing.
NEEDS_REPOS = True

# Forty aliases in one query. GitHub costs a GraphQL call by the nodes it
# returns rather than by the aliases in it, and forty repos' worth of rollup
# comes back well inside the point limit while keeping the document small
# enough to read in a log line.
CHUNK = 40

# The whole push has to finish. Two chunks at twenty-five seconds is still
# inside what a ten minute cadence can afford, and a hang is reported as a hang.
TIMEOUT_S = 25

DEFAULT_TTL_S = 600

# Alongside office-sync's own access cache, for the same reason: it is state,
# not configuration, and it must never land in the repo.
STATE = pathlib.Path.home() / ".local/state/nexus-office"


def cache_path() -> pathlib.Path:
    v = os.environ.get("OFFICE_CI_CACHE", "").strip()
    return pathlib.Path(v).expanduser() if v else STATE / "ci.json"


def ttl() -> int:
    try:
        return max(0, int(os.environ.get("OFFICE_CI_TTL_S", "").strip() or DEFAULT_TTL_S))
    except ValueError:
        return DEFAULT_TTL_S


# ── the query ────────────────────────────────────────────────────────────────

# One alias per repo. `object(expression: "HEAD:.github/workflows")` is what
# tells "no CI configured" apart from "CI configured and never ran", and those
# two have nothing in common except that neither is green.
_FRAGMENT = """
  r%d: repository(owner: %s, name: %s) {
    nameWithOwner
    defaultBranchRef {
      name
      target {
        ... on Commit {
          oid
          committedDate
          statusCheckRollup {
            state
            contexts(first: 30) {
              nodes {
                ... on CheckRun {
                  name conclusion status completedAt detailsUrl
                }
                ... on StatusContext {
                  context state targetUrl createdAt
                }
              }
            }
          }
        }
      }
    }
    workflows: object(expression: "HEAD:.github/workflows") {
      ... on Tree { entries { name } }
    }
  }
"""


def query_for(chunk: list[str]) -> str:
    parts = []
    for i, nwo in enumerate(chunk):
        owner, _, name = nwo.partition("/")
        parts.append(_FRAGMENT % (i, json.dumps(owner), json.dumps(name)))
    return "query {" + "".join(parts) + "}"


# ── reading one repo's answer ────────────────────────────────────────────────

# A check run that ended in any of these is a job somebody has to look at.
# `cancelled` is in here on purpose: a cancelled required job blocks a merge
# exactly as hard as a failed one, and calling it "not a failure" is how a
# branch sits red for a week with nobody assigned to it.
BAD_CONCLUSIONS = {"FAILURE", "TIMED_OUT", "CANCELLED", "ACTION_REQUIRED", "STARTUP_FAILURE"}
BAD_STATUSES = {"FAILURE", "ERROR"}

ROLLUP = {
    "SUCCESS": "passing",
    "FAILURE": "failing",
    "ERROR": "failing",
    "PENDING": "running",
    "EXPECTED": "running",
}


def _contexts(rollup: dict) -> list[dict]:
    nodes = ((rollup or {}).get("contexts") or {}).get("nodes") or []
    return [n for n in nodes if isinstance(n, dict) and n]


def _jobs(rollup: dict) -> tuple[list[dict], str | None]:
    """The failing jobs by name and link, plus the newest completion time seen.

    Both halves come out of the same walk because they read the same nodes, and
    the time is the honest age of the check rather than of the commit.
    """
    failing = []
    latest = None
    for n in _contexts(rollup):
        # A CheckRun carries `name`; a StatusContext carries `context`. The two
        # shapes arrive interleaved in one list and neither is optional.
        if "context" in n:
            name = str(n.get("context") or "")
            bad = str(n.get("state") or "").upper() in BAD_STATUSES
            url = n.get("targetUrl") or ""
            at = n.get("createdAt")
        else:
            name = str(n.get("name") or "")
            bad = str(n.get("conclusion") or "").upper() in BAD_CONCLUSIONS
            url = n.get("detailsUrl") or ""
            at = n.get("completedAt")
        if at and (latest is None or str(at) > latest):
            latest = str(at)
        if bad:
            failing.append({"name": name[:120], "url": str(url or "")[:400]})
    return failing[:20], latest


def _has_workflows(node: dict) -> bool:
    entries = (node or {}).get("entries")
    if not isinstance(entries, list):
        return False
    return any(str(e.get("name") or "").endswith((".yml", ".yaml")) for e in entries)


def row(nwo: str, node: dict | None, why: str = "") -> dict:
    """One desk's line, from one repository node. `node` None means we could not look."""
    base = {"repo": nwo, "ci": "unknown", "branch": "", "detail": why[:200],
            "failing": [], "checked_at": None, "commit_at": None, "run_url": ""}
    if not node:
        base["detail"] = why[:200] or "the repository could not be read with the token we hold"
        return base

    ref = node.get("defaultBranchRef") or {}
    base["branch"] = str(ref.get("name") or "")
    target = ref.get("target") or {}
    base["commit_at"] = target.get("committedDate")

    rollup = target.get("statusCheckRollup")
    workflows = _has_workflows(node.get("workflows"))

    if not ref:
        base["detail"] = "the repository has no default branch, so there is nothing to build"
        base["ci"] = "none"
        return base

    if not rollup:
        base["ci"] = "never" if workflows else "none"
        base["detail"] = ("there are workflow files, and the head of the default branch "
                          "has no check on it at all"
                          if workflows else "no workflow files, so nothing here builds anything")
        return base

    failing, checked_at = _jobs(rollup)
    base["failing"] = failing
    base["checked_at"] = checked_at
    base["ci"] = ROLLUP.get(str(rollup.get("state") or "").upper(), "unknown")
    if base["ci"] == "unknown":
        base["detail"] = f"the check rollup said {rollup.get('state')!r}, which nothing here knows"
    if failing:
        base["run_url"] = failing[0]["url"]
        base["detail"] = (f"{len(failing)} job{'' if len(failing) == 1 else 's'} "
                          f"failing on {base['branch'] or 'the default branch'}")
    return base


# ── asking GitHub ────────────────────────────────────────────────────────────

def _gh(chunk: list[str]) -> tuple[dict, str]:
    """One GraphQL call. Returns (data-by-alias, error). Both can be non-empty:
    GraphQL answers a partly-unreadable query with data AND errors, and throwing
    the readable half away because one repo is private would be silly."""
    try:
        proc = subprocess.run(
            ["gh", "api", "graphql", "-f", "query=" + query_for(chunk)],
            capture_output=True, text=True, timeout=TIMEOUT_S,
        )
    except subprocess.TimeoutExpired:
        return {}, f"gh did not answer in {TIMEOUT_S}s"
    except FileNotFoundError:
        return {}, "there is no gh on this machine to ask"
    except OSError as exc:
        return {}, f"{type(exc).__name__}: {exc}"[:200]

    try:
        body = json.loads(proc.stdout or "")
    except json.JSONDecodeError:
        detail = (proc.stderr or proc.stdout or "").strip().replace("\n", " ")[:200]
        return {}, detail or "gh printed nothing that parsed"

    data = body.get("data") or {}
    errs = body.get("errors") or []
    msg = "; ".join(str(e.get("message") or "")[:120] for e in errs[:3])
    if not data and not msg:
        msg = "gh answered with neither data nor an error"
    return data, msg


def fetch(repos: list[str]) -> tuple[list[dict], str]:
    rows = []
    problems = []
    for i in range(0, len(repos), CHUNK):
        chunk = repos[i:i + CHUNK]
        data, err = _gh(chunk)
        if err:
            problems.append(err)
        if not data:
            # The whole chunk is unreadable. Every desk in it says so, rather
            # than quietly not appearing on the board.
            rows.extend(row(nwo, None, err) for nwo in chunk)
            continue
        for n, nwo in enumerate(chunk):
            rows.append(row(nwo, data.get(f"r{n}"), err))
    return rows, "; ".join(problems)[:300]


# ── the cache ────────────────────────────────────────────────────────────────

def load_cache() -> dict | None:
    try:
        body = json.loads(cache_path().read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    return body if isinstance(body, dict) and isinstance(body.get("rows"), list) else None


def save_cache(body: dict) -> None:
    try:
        p = cache_path()
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(json.dumps(body), encoding="utf-8")
    except OSError:
        # A cache that cannot be written costs an API call, not a snapshot.
        pass


def _iso(epoch: float) -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(epoch))


# ── the section ──────────────────────────────────────────────────────────────

STATES = ("failing", "never", "running", "passing", "none", "unknown")

# Faults first, then the ones nobody has to do anything about. `none` is not an
# alarm and never leads.
ORDER = {"failing": 0, "unknown": 1, "never": 2, "running": 3, "passing": 4, "none": 5}


def summarise(rows: list[dict], fetched: float, state: str, detail: str) -> dict:
    counts = {k: 0 for k in STATES}
    for r in rows:
        counts[r["ci"]] = counts.get(r["ci"], 0) + 1
    rows = sorted(rows, key=lambda r: (ORDER.get(r["ci"], 9), r["repo"]))
    return {
        "state": state,
        "detail": detail[:300],
        "checked": len(rows),
        "counts": counts,
        # `unknown` counts as an alarm. Not looking is not the same as being
        # fine, and only one of those two deserves a quiet room.
        "alarm": counts["failing"] + counts["never"] + counts["unknown"],
        "fetched_at": _iso(fetched),
        "age_s": max(0, int(time.time() - fetched)),
        "ttl_s": ttl(),
        "repos": rows,
    }


def read(repos=()) -> dict:
    repos = sorted({r for r in (repos or []) if isinstance(r, str) and "/" in r})
    if not repos:
        return {"state": "no-desks",
                "detail": "there are no repos in the snapshot, so there is no build to watch"}

    cached = load_cache()
    now = time.time()
    if cached and now - float(cached.get("at") or 0) < ttl() \
            and sorted(r["repo"] for r in cached["rows"]) == repos:
        return summarise(cached["rows"], float(cached["at"]), "ok", "")

    rows, err = fetch(repos)
    unreadable = all(r["ci"] == "unknown" for r in rows)
    if unreadable and cached:
        # The refresh failed outright and we still hold a board. Showing the old
        # one WITH its age and the reason beats showing nothing: the age is
        # already on every row, so nobody can mistake it for fresh.
        return summarise(cached["rows"], float(cached.get("at") or 0), "cached", err)
    if unreadable:
        return {"state": "unreadable",
                "detail": err[:300] or "nothing could be read about any repo"}

    save_cache({"at": now, "rows": rows})
    return summarise(rows, now, "ok", err)

File: src/scene/fixtures/ci.js
/**
 * the build lamps
 *
 * A lamp over every desk saying whether that repo's default branch actually
 * builds. The runner opens PRs into these repos all day, so a red default
 * branch is the most expensive thing in the room not to notice: nothing stops,
 * the PRs keep stacking, and every one of them is built on a broken base.
 *
 * WHY IT HANGS AT 2.7 AND NOT ON THE DESK
 * The desk's pick pad is a 2.4m-tall box centred at y 1.2, and it ENCLOSES
 * everything on the desk top. The in-tray gave up its click target to that box
 * and says so in its own file. This lamp has to be clickable — the issue asks
 * for the failing job and a link to the run — so it hangs above the pad, just
 * over head height, where a ray reaches it before anything else. No change to
 * office.js was needed for that, which is the point of the height.
 *
 * SIX STATES, NOT TWO
 * Passing and failing are the easy pair. `never` is a repo with workflow files
 * whose head commit carries no check at all, which is usually a workflow
 * filtered out of the branch it was meant to run on — a different afternoon
 * from a build that broke. `none` is a repo with no CI, which is a DECISION and
 * is drawn as a quiet grey ring rather than a fault. `running` is a rollup
 * still in flight, and calling it either colour would be a lie that resolves
 * itself in four minutes. `unknown` is "we could not look", which must never
 * render as "it is fine".
 */

import * as THREE from "three";
import * as BGU from "three/examples/jsm/utils/BufferGeometryUtils.js";
import { toon } from "../kit.js";

export const id = "ci";
export const title = "the build lamps";
export const wall = false;

/**
 * One glyph and one colour per state, and one plain-language name.
 *
 * `size` is the whole loudness budget, and the first version of this file spent
 * it wrong: every lamp was a cream bubble with a small coloured glyph, so the
 * picture of the room showed twelve identical white dots and the failing build
 * — the entire reason this fixture exists — was invisible from across the
 * floor. So an alarm lamp is now a SOLID disc in its own colour with a white
 * glyph on it, and a quiet one stays cream. Colour does the shouting; the glyph
 * is only there for whoever cannot separate the two reds from the greens.
 *
 * A passing lamp is small and quiet because seventy shouting green ticks is a
 * room you stop reading. `none` is smaller still and grey: drawn, so a repo with
 * no CI is never confused with a repo this fixture forgot, and never loud, so a
 * deliberate absence of CI is never mistaken for a fault.
 */
const STATE = {
  failing: { glyph: "✕", dot: "#d1495b", name: "failing", size: 1.55, alarm: true },
  unknown: { glyph: "?", dot: "#d9a441", name: "could not look", size: 1.05, alarm: true },
  never: { glyph: "–", dot: "#7b5ea7", name: "never checked", size: 1.0, alarm: true },
  running: { glyph: "●", dot: "#5b8dd9", name: "running now", size: 0.62, alarm: false },
  passing: { glyph: "✓", dot: "#3f9e6a", name: "passing", size: 0.52, alarm: false },
  none: { glyph: "○", dot: "#a9a29b", name: "no CI at all", size: 0.44, alarm: false },
};
const spec = (s) => STATE[s] || STATE.unknown;

const CREAM = "#fffdf5";

/** Why the source could not answer, said out loud rather than drawn as green. */
const BROKEN = {
  "no-desks": "there are no desks, so there is nothing whose build to watch",
  unreadable: "nothing could be read about any repo's build",
  error: "reading the build checks failed",
  unbuilt: "this source has not been built yet",
};

/** Above the desk's 2.4m pick pad, and above a seated villager's head. */
const LAMP_Y = 2.72;

/**
 * The back right corner of the desk top, and not the middle of it.
 *
 * Dead centre was tried and photographed: the pole ran straight down the face
 * of the villager at the desk BEHIND, which is this project's oldest defect
 * shape and the reason the name plaque was moved once already. The desk top is
 * 1.7 x 0.85 centred on the station, the monitor foot sits at z -0.2 and the
 * mug at x +0.58, so this corner is the one piece of wood nothing else wants.
 */
const FLAG_X = 0.72;
const FLAG_Z = -0.34;
const TOP = 0.78;

/* ---------------------------------------------------------------- the build -- */

const S = 192;

/**
 * The disc. An alarm is filled with its own colour and haloed, a quiet state is
 * cream with a coloured rim, and the halo is what makes a red lamp survive being
 * twenty-six pixels wide against a tan floor.
 */
function texture(state) {
  const sp = spec(state);
  const c = document.createElement("canvas");
  c.width = c.height = S;
  const g = c.getContext("2d");
  const cx = S / 2;

  if (sp.alarm) {
    const halo = g.createRadialGradient(cx, cx, S * 0.3, cx, cx, S * 0.5);
    halo.addColorStop(0, `${sp.dot}66`);
    halo.addColorStop(1, `${sp.dot}00`);
    g.fillStyle = halo;
    g.fillRect(0, 0, S, S);
  }

  g.fillStyle = sp.alarm ? sp.dot : CREAM;
  g.beginPath();
  g.arc(cx, cx, S * 0.33, 0, Math.PI * 2);
  g.fill();
  g.strokeStyle = sp.alarm ? "rgba(255,255,255,0.85)" : sp.dot;
  g.lineWidth = S * (sp.alarm ? 0.03 : 0.05);
  g.stroke();

  g.fillStyle = sp.alarm ? "#fff" : sp.dot;
  g.font = `800 ${Math.round(S * 0.36)}px ui-rounded, "SF Pro Rounded", Quicksand, system-ui, sans-serif`;
  g.textAlign = "center";
  g.textBaseline = "middle";
  g.fillText(sp.glyph, cx, cx + S * 0.02);

  const t = new THREE.CanvasTexture(c);
  t.colorSpace = THREE.SRGBColorSpace;
  return t;
}

// One sprite per state, cloned per desk. Sprite.clone() shares the material and
// therefore the canvas texture, so seventy lamps cost six textures.
const protos = new Map();
function lamp(state) {
  if (!protos.has(state)) {
    const s = new THREE.Sprite(new THREE.SpriteMaterial({
      map: texture(state), depthTest: false, transparent: true,
    }));
    s.scale.setScalar(0.62 * spec(state).size);
    // Over the walls and the furniture, under the banners that carry words.
    s.renderOrder = 9;
    protos.set(state, s);
  }
  return protos.get(state).clone();
}

export function build(ctx) {
  const sec = ctx.section;
  const stations = ctx.stations || [];
  if (!sec || sec.state === "no-desks" || !stations.length) return null;

  // The source is broken, so every desk gets the amber "we did not look" lamp
  // rather than no lamp. A room that draws nothing when its watcher dies is a
  // room that reports every broken build as green.
  const byRepo = new Map();
  for (const r of sec.repos || []) byRepo.set(r.repo, r);

  // Loudest first. Not for the drawing — a sprite's order does not change what
  // it looks like — but because the shot harness clicks the FIRST pickable a
  // fixture offers, and a picture of a passing desk is not a picture of what
  // this fixture is for.
  const order = ["failing", "unknown", "never", "running", "passing", "none"];
  const byLoudness = [...stations].sort((a, b) =>
    order.indexOf(byRepo.get(a.repo)?.ci ?? "unknown") -
    order.indexOf(byRepo.get(b.repo)?.ci ?? "unknown"));

  const group = new THREE.Group();

  // A pole, for the alarm lamps only.
  //
  // The lamp hangs level with the row behind it, so in the room shot a red
  // cross floats midway between two desks and you have to count to work out
  // which one it belongs to. A pole rising out of the desk answers that
  // instantly. Only faults get one: seventy poles for seventy green ticks is
  // the clutter this fixture is supposed to save you from.
  const poles = new Map();
  const POLE_H = LAMP_Y - 0.2 - TOP;
  const poleGeo = new THREE.CylinderGeometry(0.022, 0.022, POLE_H, 6);

  for (const st of byLoudness) {
    const r = byRepo.get(st.repo) || {
      repo: st.repo, ci: "unknown", failing: [], branch: "",
      detail: BROKEN[sec.state] || `the build source said "${sec.state}"`,
    };
    const s = lamp(r.ci);
    s.position.set(st.x + FLAG_X, LAMP_Y, st.z + FLAG_Z);
    // Every lamp is clickable, not only the red ones. The age of a green check
    // is the thing that tells a fresh pass from a stale one, and a state you
    // cannot click is a state you have to take on trust.
    s.userData.fixture = { id, payload: { ...sec, focus: r.repo } };
    group.add(s);

    const sp = spec(r.ci);
    if (sp.alarm) {
      if (!poles.has(sp.dot)) poles.set(sp.dot, []);
      poles.get(sp.dot).push(poleGeo.clone().applyMatrix4(
        new THREE.Matrix4().makeTranslation(st.x + FLAG_X, TOP + POLE_H / 2, st.z + FLAG_Z)
      ));
    }
  }
  for (const [colour, geos] of poles) {
    group.add(new THREE.Mesh(BGU.mergeGeometries(geos), toon(colour)));
  }
  return group.children.length ? group : null;
}

/* ---------------------------------------------------------------- the panel -- */

const relative = (iso) => {
  if (!iso) return null;
  const mins = Math.round((Date.now() - Date.parse(iso)) / 60000);
  if (!Number.isFinite(mins)) return null;
  if (mins < 1) return "just now";
  if (mins < 60) return `${mins}m ago`;
  if (mins < 60 * 36) return `${Math.round(mins / 60)}h ago`;
  return `${Math.round(mins / 1440)}d ago`;
};

function card(el, r) {
  const sp = spec(r.ci);
  const box = el("div", `issue${sp.alarm ? " hot" : ""}`);

  const top = el("div", "issue-top");
  top.append(el("span", "issue-title", r.repo));
  box.append(top);

  const tags = el("div", "tags");
  const state = el("span", `tag${sp.alarm ? " hot" : ""}`, sp.name);
  if (!sp.alarm) state.style.color = sp.dot;
  tags.append(state);
  if (r.branch) tags.append(el("span", "tag", r.branch));
  box.append(tags);

  if (r.detail) box.append(el("p", "p-detail", r.detail));

  // The age of the check, from the check runs themselves. When none of them
  // reported a time it says so: the commit's date is beside it, labelled as the
  // commit's date, and never quietly promoted into a check time.
  const line = el("p", "log");
  const checked = relative(r.checked_at);
  line.append(el("b", null, "checked"), " ");
  line.append(el("span", checked ? "" : "wait",
    checked || "no check reported a time"));
  const committed = relative(r.commit_at);
  if (committed) line.append(" · ", el("b", null, "head commit"), ` ${committed}`);
  box.append(line);

  for (const j of r.failing || []) {
    const row = el("p", "log");
    row.append(el("span", "bad", j.name || "an unnamed job"));
    if (j.url) {
      row.append(" ");
      const a = el("a", "link", "open the run");
      a.href = j.url;
      a.target = "_blank";
      a.rel = "noopener noreferrer";
      row.append(a);
    }
    box.append(row);
  }
  return box;
}

export function panel(payload, world, api) {
  const { el } = api;
  const wrap = el("div");
  const sec = payload || world?.sections?.ci || { state: "unbuilt" };

  if (sec.state !== "ok" && sec.state !== "cached") {
    wrap.append(el("p", "empty", BROKEN[sec.state] || `the build source said "${sec.state}"`));
    if (sec.detail) wrap.append(el("pre", "gate-target", sec.detail));
    return wrap;
  }

  const age = relative(sec.fetched_at);
  const sum = el("p", "log");
  sum.append(el("b", null, `${sec.checked} repos`), " · ");
  sum.append("asked GitHub ", el("b", null, age || "at an unknown time"));
  wrap.append(sum);

  // A cached board is the honest answer when the refresh failed, and it is only
  // honest while it is labelled. The age above is what makes it readable.
  if (sec.state === "cached") {
    wrap.append(el("p", "gate-clock",
      "this is the last board that came back. The refresh failed, so nothing " +
      "here has been checked since the time above."));
  }
  if (sec.detail) wrap.append(el("pre", "gate-target", sec.detail));

  const c = sec.counts || {};
  const counts = el("p", "log");
  for (const k of ["failing", "unknown", "never", "running", "passing", "none"]) {
    if (!c[k]) continue;
    counts.append(el("span", spec(k).alarm ? "bad" : k === "passing" ? "ok" : "",
      `${c[k]} ${spec(k).name}`), "  ");
  }
  wrap.append(counts);

  const rows = sec.repos || [];
  // The desk you clicked goes first, whatever its state. Clicking one lamp and
  // being handed a list to search is a click that did not answer anything.
  const focused = rows.find((r) => r.repo === sec.focus);
  if (focused) {
    wrap.append(el("h3", null, "the desk you clicked"));
    wrap.append(card(el, focused));
  }

  const groups = [
    ["failing", rows.filter((r) => r.ci === "failing" && r !== focused)],
    ["could not look", rows.filter((r) => r.ci === "unknown" && r !== focused)],
    ["never checked", rows.filter((r) => r.ci === "never" && r !== focused)],
    ["running now", rows.filter((r) => r.ci === "running" && r !== focused)],
    ["passing", rows.filter((r) => r.ci === "passing" && r !== focused)],
    ["no CI at all, on purpose", rows.filter((r) => r.ci === "none" && r !== focused)],
  ];
  for (const [label, list] of groups) {
    if (!list.length) continue;
    wrap.append(el("h3", null, `${label} (${list.length})`));
    for (const r of list) wrap.append(card(el, r));
  }
  if (!rows.length) {
    wrap.append(el("p", "empty", "GitHub answered, and no desk here has a repo to build."));
  }
  return wrap;
}

/* ----------------------------------------------------------------- the demo -- */

const iso = (h) => new Date(Date.now() - h * 3600000).toISOString().replace(/\.\d+Z$/, "Z");

/**
 * Every state, on purpose, against the demo floor's own repos. A demo that only
 * shows green lamps is a demo that lets the red one rot untested, and `none` and
 * `unknown` are the two most likely to be quietly got wrong.
 */
export function demo() {
  const r = (repo, ci, o = {}) => ({
    repo, ci, branch: "main", detail: "", failing: [], run_url: "",
    checked_at: iso(0.6), commit_at: iso(0.7), ...o,
  });
  const repos = [
    r("acme/storefront", "failing", {
      detail: "2 jobs failing on main",
      run_url: "https://github.com/acme/storefront/actions/runs/9001",
      failing: [
        { name: "test (node 22)", url: "https://github.com/acme/storefront/actions/runs/9001" },
        { name: "typecheck", url: "https://github.com/acme/storefront/actions/runs/9001" },
      ],
    }),
    r("northwind/api", "failing", {
      detail: "1 job failing on main", checked_at: iso(31), commit_at: iso(31),
      run_url: "https://github.com/northwind/api/actions/runs/4412",
      failing: [{ name: "integration", url: "https://github.com/northwind/api/actions/runs/4412" }],
    }),
    r("acme/mobile", "never", {
      detail: "there are workflow files, and the head of the default branch has no check on it at all",
      checked_at: null, commit_at: iso(3),
    }),
    r("northwind/warehouse", "running", { detail: "", checked_at: null, commit_at: iso(0.2) }),
    r("acme/website", "passing"),
    r("acme/billing", "passing", { checked_at: iso(90), commit_at: iso(90) }),
    r("northwind/analytics", "passing", { checked_at: iso(5) }),
    r("tiny/experiments", "passing", { checked_at: iso(200), commit_at: iso(200) }),
    r("acme/docs", "none", {
      detail: "no workflow files, so nothing here builds anything",
      checked_at: null, commit_at: iso(48),
    }),
    r("acme/legacy-import", "none", {
      detail: "no workflow files, so nothing here builds anything",
      checked_at: null, commit_at: iso(900),
    }),
    r("tiny/scratch", "none", {
      detail: "no workflow files, so nothing here builds anything",
      checked_at: null, commit_at: iso(1400),
    }),
    r("northwind/internal-tools", "unknown", {
      branch: "",
      detail: "the repository could not be read with the token we hold",
      checked_at: null, commit_at: null,
    }),
  ];
  const counts = { failing: 0, unknown: 0, never: 0, running: 0, passing: 0, none: 0 };
  for (const x of repos) counts[x.ci]++;
  return {
    state: "ok",
    detail: "",
    checked: repos.length,
    counts,
    alarm: counts.failing + counts.never + counts.unknown,
    fetched_at: iso(0.05),
    age_s: 180,
    ttl_s: 600,
    repos,
  };
}
